File: Extra/Frontier Scientist Related Files/iOS-7-31-14-/FSDemo/ArticlesTableViewController.m
//
//  ArticlesTableViewController.m
//  FSDemo
//
//  Created by alandrews3 on 5/15/14.
//  Copyright (c) 2014 Andrew Clark. All rights reserved.
//

#import "ArticlesTableViewController.h"

@interface ArticlesTableViewController ()

@end

@implementation ArticlesTableViewController

@synthesize progressIndicator;
@synthesize storyLink;
@synthesize storyIndex;

- (id)initWithStyle:(UITableViewStyle)style
{
    self = [super initWithStyle:style];
    if (self) {
        // Custom initialization
    }
    return self;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    // Uncomment the following line to preserve selection between presentations.
    // self.clearsSelectionOnViewWillAppear = NO;
    
    // Uncomment the following line to display an Edit button in the navigation bar for this view controller.
    // self.navigationItem.rightBarButtonItem = self.editButtonItem;
    self.view.backgroundColor = [UIColor colorWithPatternImage:[UIImage imageNamed:@"RiR_lined_iphone_V44px"]];
    UIImageView *backgroundView = [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"RiR_lined_iphone_V44px"]];
    [self.view addSubview:backgroundView];
    [self.view sendSubviewToBack:backgroundView];
    
    feed = @"http://frontierscientists.com/feed";
    [self parseXMLFileAtURL];
}

-(void)receiveTableNotification:(NSNotification *)pNotification
{
    NSLog(@"Reload Articles time!");
    [self parseXMLFileAtURL];
    [self.tableView reloadData];
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark - Table view data source
- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    // Return the number of sections.
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    // Return the number of rows in the section.
    return [stories count];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    static NSString *CellIdentifier = @"ArticleTitle";
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:CellIdentifier forIndexPath:indexPath];
    
    cell.textLabel.text = [[stories objectAtIndex: indexPath.row] objectForKey: @"title"];
    cell.textLabel.font = [UIFont fontWithName:@"Chalkduster" size:17.0];
    
    return cell;
}


- (UIView *)tableView:(UITableView *)tableView viewForHeaderInSection:(NSInteger)section
{
    UIView *customView = [[UIView alloc] initWithFrame:CGRectMake(10, 0, 300, 2)];
    UILabel *headerLabel =  [[UILabel alloc] initWithFrame:CGRectMake(10, 0, 300, 12)];
    headerLabel.backgroundColor = [UIColor clearColor];
    
    [customView addSubview:headerLabel];
    return customView;
}

-(CGFloat)tableView:(UITableView *)tableView heightForHeaderInSection:(NSInteger)section
{
    return 15;
}


 - (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section
 {
 //PLACEHOLDER
 return @"RSS Feed";
 }

 //This function is from the old app
- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
	storyLink = [[stories objectAtIndex: indexPath.row] objectForKey: @"link"];
	
	// clean up the link - get rid of spaces, returns, and tabs...
	storyLink = [storyLink stringByReplacingOccurrencesOfString:@" " withString:@""];
	storyLink = [storyLink stringByReplacingOccurrencesOfString:@"\n" withString:@""];
	storyLink = [storyLink stringByReplacingOccurrencesOfString:@"	" withString:@""];
	
    [self performSegueWithIdentifier:@"RSSFeed" sender:indexPath];

}

- (UIActivityIndicatorView *)progressIndicator
{
    if (progressIndicator == nil)
    {
        CGRect frame = CGRectMake(self.view.frame.size.width/2-15, self.view.frame.size.height/2-15, 30, 30);
        progressIndicator = [[UIActivityIndicatorView alloc] initWithFrame:frame];
        [progressIndicator startAnimating];
        progressIndicator.activityIndicatorViewStyle = UIActivityIndicatorViewStyleGray;
        [progressIndicator sizeToFit];
        progressIndicator.autoresizingMask = (UIViewAutoresizingFlexibleLeftMargin |
                                              UIViewAutoresizingFlexibleRightMargin |
                                              UIViewAutoresizingFlexibleTopMargin |
                                              UIViewAutoresizingFlexibleBottomMargin);
        
        progressIndicator.tag = 1;    // tag this view for later so we can remove it from recycled table cells
    }
    return progressIndicator;
}


- (void)parserDidStartDocument:(NSXMLParser *)parser{
	NSLog(@"found file and started parsing");
	
}
//- (void)parseXMLFileAtURL:(NSString *)URL
- (void)parseXMLFileAtURL
{
	stories = [[NSMutableArray alloc] init];
	NSString *URL = feed;
    //you must then convert the path to a proper NSURL or it won't work
    NSURL *xmlURL = [NSURL URLWithString:URL];
	
    // here, for some reason you have to use NSClassFromString when trying to alloc NSXMLParser, otherwise you will get an object not found error
    // this may be necessary only for the toolchain
    rssParser = [[NSXMLParser alloc] initWithContentsOfURL:xmlURL];
	
    // Set self as the delegate of the parser so that it will receive the parser delegate methods callbacks.
    [rssParser setDelegate:self];
	
    // Depending on the XML document you're parsing, you may want to enable these features of NSXMLParser.
    [rssParser setShouldProcessNamespaces:NO];
    [rssParser setShouldReportNamespacePrefixes:NO];
    [rssParser setShouldResolveExternalEntities:NO];
	
    [rssParser parse];
	
}

- (void)parser:(NSXMLParser *)parser parseErrorOccurred:(NSError *)parseError {
	NSString * errorString = [NSString stringWithFormat:@"Unable to download story feed from web site (Error code %li )", (long)[parseError code]];
	NSLog(@"error parsing XML: %@", errorString);
	
	UIAlertView * errorAlert = [[UIAlertView alloc] initWithTitle:@"Error loading content" message:errorString delegate:self cancelButtonTitle:@"OK" otherButtonTitles:nil];
	[errorAlert show];
}

- (void)parser:(NSXMLParser *)parser didStartElement:(NSString *)elementName namespaceURI:(NSString *)namespaceURI qualifiedName:(NSString *)qName attributes:(NSDictionary *)attributeDict{
    NSLog(@"found this element: %@", elementName);
	currentElement = [elementName copy];
	if ([elementName isEqualToString:@"item"]) {
		// clear out our story item caches...
		item = [[NSMutableDictionary alloc] init];
		currentTitle = [[NSMutableString alloc] init];
		currentDate = [[NSMutableString alloc] init];
		currentSummary = [[NSMutableString alloc] init];
		currentLink = [[NSMutableString alloc] init];
	}
	
}

- (void)parser:(NSXMLParser *)parser didEndElement:(NSString *)elementName namespaceURI:(NSString *)namespaceURI qualifiedName:(NSString *)qName{
	//NSLog(@"ended element: %@", elementName);
	if ([elementName isEqualToString:@"item"]) {
		// save values to an item, then store that item into the array...
		[item setObject:currentTitle forKey:@"title"];
		[item setObject:currentLink forKey:@"link"];
		[item setObject:currentSummary forKey:@"summary"];
		[item setObject:currentDate forKey:@"date"];
		
		[stories addObject:[item copy]];
		NSLog(@"adding story: %@", currentTitle);
        
        }
}

- (void)parser:(NSXMLParser *)parser foundCharacters:(NSString *)string{
	//NSLog(@"found characters: %@", string);
	// save the characters for the current item...
	if ([currentElement isEqualToString:@"title"]) {
		[currentTitle appendString:string];
	} else if ([currentElement isEqualToString:@"link"]) {
		[currentLink appendString:string];
	} else if ([currentElement isEqualToString:@"description"]) {
		[currentSummary appendString:string];
	} else if ([currentElement isEqualToString:@"pubDate"]) {
		[currentDate appendString:string];
	}
	
}

- (void)parserDidEndDocument:(NSXMLParser *)parser {
	
	[activityIndicator stopAnimating];
	[activityIndicator removeFromSuperview];
	
	NSLog(@"all done!");
	NSLog(@"stories array has %lu items", (unsigned long)[stories count]);
    
}

/*
// Override to support conditional editing of the table view.
- (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath
{
    // Return NO if you do not want the specified item to be editable.
    return YES;
}
*/

/*
// Override to support editing the table view.
- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath
{
    if (editingStyle == UITableViewCellEditingStyleDelete) {
        // Delete the row from the data source
        [tableView deleteRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationFade];
    } else if (editingStyle == UITableViewCellEditingStyleInsert) {
        // Create a new instance of the appropriate class, insert it into the array, and add a new row to the table view
    }   
}
*/

/*
// Override to support rearranging the table view.
- (void)tableView:(UITableView *)tableView moveRowAtIndexPath:(NSIndexPath *)fromIndexPath toIndexPath:(NSIndexPath *)toIndexPath
{
}
*/

/*
// Override to support conditional rearranging of the table view.
- (BOOL)tableView:(UITableView *)tableView canMoveRowAtIndexPath:(NSIndexPath *)indexPath
{
    // Return NO if you do not want the item to be re-orderable.
    return YES;
}
*/


#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
    
    if([[segue identifier] isEqualToString:@"RSSFeed"])
    {
        ArticlesViewController *avc = [segue destinationViewController];
        avc.urlString = storyLink;
    }
    
}


@end

File: FrontierScientistsMedia/FrontierScientistsMedia/ArticlesTableViewController.swift
// ArticlesTableViewController.swift

import Foundation
import UIKit

/*
    This is the ArticlesTableViewController class, responsable for parsing the articles feed and displaying
    the articles in a TableView.
*/
class ArticlesTableViewController: UITableViewController, NSXMLParserDelegate{
    
/*
    Outlets
*/
    @IBOutlet var articleTableView: UITableView?
/*
    Class Constants
*/
    let feed: String = "http://frontierscientists.com/feed"
/*
    Class Variables
*/
    var articleLink: String = ""
    var articleReadStatusFileDict: NSMutableDictionary = NSMutableDictionary()
    var articleReadStatusFilePath: String = NSHomeDirectory() + "Library/Caches/ArticlesReadStatus.txt"
    var articleTitle: NSMutableString = NSMutableString()
    var element: String = ""
    var parser = NSXMLParser()
    var posts = [[String(): String()]]
    var selectedArtcileIndex: Int = 0
    
/*
    Class Functions
*/
    // viewDidLoad
    override func viewDidLoad() {
        super.viewDidLoad()
        
        // Setup
        self.view.backgroundColor = UIColor(patternImage:UIImage(named: "bg.png")!)
        self.title = "Articles"
        
        beginParsing()
        articleReadStatusFileDict = [NSString(string: "dummy"): NSNumber(int: 0)]
        if (!NSFileManager.defaultManager().fileExistsAtPath(articleReadStatusFilePath)) {
            createReadArticlesFileIfNone()
        } else {
            loadArticlesReadStatusFileToArticlesReadStatusDict()
        }
        updateArticlesReadStatusDict()
        saveArticlesReadStatusDictToArticlesReadStatusFile()
        print("Finished loading articles.")
    }
    // prepareForSegue
    override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject?) {
        if (segue.identifier == "RSSFeed") {
            let avc = segue.destinationViewController as? ArticleViewController
            var str = posts[selectedArtcileIndex]["link"]
            // Ready link for use in WebView
            str = str?.stringByReplacingOccurrencesOfString(" ", withString: "")
            str = str?.stringByReplacingOccurrencesOfString("\n", withString: "")
            str = str?.stringByReplacingOccurrencesOfString("   ", withString: "")
            str = str?.stringByReplacingOccurrencesOfString("\t", withString: "")
            // Set link string
            avc?.articleLinkString = str
        }
    }
    
/*
    TableView Functions
*/
    // numberOfSectionsInTableView
    override func numberOfSectionsInTableView(tableView: UITableView) -> Int {
        return 1
    }
    // heightForHeaderInSection
    override func tableView(tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
        return 0.001
    }
    // numberOfRowsInSection
    override func tableView(tableView: UITableView, numberOfRowsInSection section: Int) ->Int {
        return 10
    }
    // heightForRowAtIndexPath
    override func tableView(tableView: UITableView, heightForRowAtIndexPath indexPath: NSIndexPath) -> CGFloat {
        if (UIDevice.currentDevice().userInterfaceIdiom == iPadDeviceType) {
            return 80
        } else { // iPhone
            return 40
        }
    }
    // cellForRowAtIndexPath
    override func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        // Initialize cell
        let cell: UITableViewCell = UITableViewCell(style: UITableViewCellStyle.Default, reuseIdentifier: "ArticleTitle")
        // Format cell
        cell.backgroundColor = UIColor.clearColor()
        cell.backgroundView = UIImageView(image: UIImage(named: "CellBorder.png"))
        cell.textLabel?.backgroundColor = UIColor.clearColor()
        cell.textLabel!.text = posts[indexPath.row]["title"]!
        cell.contentView.layer.borderColor = UIColor.clearColor().CGColor
        // Include "new" icon only when article is unread
        var accessoryImageView: UIImageView
        if (articleHasBeenRead(indexPath.row)) {
            accessoryImageView = UIImageView(image: UIImage(named: "Transition_Icon.png"))
        } else {
            accessoryImageView = UIImageView(image: UIImage(named: "new_icon-web.png"))
        }
        // Change image text sizes based on screen size
        if (UIDevice.currentDevice().userInterfaceIdiom == iPadDeviceType) {
            accessoryImageView.frame = CGRect(x: 0, y: 0, width: 80, height: 80)
            cell.textLabel?.font = UIFont(name: "Chalkduster", size: 30)
        } else { // (UIDevice.currentDevice().userInterfaceIdiom.rawValue == iPhoneDeviceType)
            accessoryImageView.frame = CGRect(x: 0, y: 0, width: 40, height: 40)
            cell.textLabel?.font = UIFont(name: "Chalkduster", size: 20)
        }
        cell.accessoryView = accessoryImageView
        
        return cell
    }
    // didSelectRowAtIndexPath
    override func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath) {
        selectedArtcileIndex = indexPath.row
        let articleStoryLink = posts[indexPath.row]["link"]!
        articleReadStatusFileDict[articleStoryLink] = NSNumber(int: 1)
        saveArticlesReadStatusDictToArticlesReadStatusFile()
        self.tableView.reloadData()
        performSegueWithIdentifier("RSSFeed", sender: self)
    }
    
/*
    Helper and Content Functions
*/
    // createReadArticlesFileIfNone
    // This function creates a new "read articles" file if there is not one already present on the device.
    func createReadArticlesFileIfNone() {
        let fileMgr = NSFileManager.defaultManager()
        if (!fileMgr.fileExistsAtPath(articleReadStatusFilePath)) {
            if (!fileMgr.createFileAtPath(articleReadStatusFilePath, contents: nil, attributes: articleReadStatusFileDict as [String : AnyObject])) {
                print("Error! Creating articleReadStatusFile failed to create at \(articleReadStatusFilePath)")
            }
        }
    }
    // loadArticlesReadStatusFileToArticlesReadStatusDict
    // This function loads the contents of the "read articles" file in to the articleReadStatusFileDict
    func loadArticlesReadStatusFileToArticlesReadStatusDict() {
        articleReadStatusFileDict = NSMutableDictionary(contentsOfFile: articleReadStatusFilePath)!
    }
    // saveArticlesReadStatusDictToArticlesReadStatusFile
    // This function saves the articleReadStatusFileDict data back into a file on the device
    func saveArticlesReadStatusDictToArticlesReadStatusFile() {
        articleReadStatusFileDict.writeToFile(articleReadStatusFilePath, atomically: true)
    }
    // updateArticlesReadStatusDict
    // This function updates the articleReadStatusFileDict after a new article has been viewed
    func updateArticlesReadStatusDict() {
        var newArticlesReadStatusDict: NSMutableDictionary = NSMutableDictionary()
        let dictHasBeenReadValue: NSNumber = 1
        var articleStoryLink: NSString = ""
        newArticlesReadStatusDict = [:]
        for ii in 0...posts.count-1 {
            articleStoryLink = posts[ii]["link"]!
            
            if (articleReadStatusFileDict.objectForKey(articleStoryLink) != nil &&
                articleReadStatusFileDict.objectForKey(articleStoryLink)!.integerValue == dictHasBeenReadValue.integerValue) {
                newArticlesReadStatusDict.setObject(NSNumber(int: 1), forKey: articleStoryLink)
            } else {
                newArticlesReadStatusDict.setObject(NSNumber(int: 0), forKey: articleStoryLink)
            }
        }
        articleReadStatusFileDict = newArticlesReadStatusDict
    }
    // articleHasBeenRead
    // This function returns a bool based on whether or not the article at the passed index has been viewed
    func articleHasBeenRead(index: Int) -> Bool {
        let linkString: String = posts[index]["link"]!
        if (articleReadStatusFileDict.objectForKey(linkString) != nil &&
            articleReadStatusFileDict.objectForKey(linkString)?.integerValue == NSNumber(int: 1)) {
            return true
        }
        return false
    }
    
/*
    Parser Functions
*/
    // parseErrorOccurred
    func parser(parser: NSXMLParser, parseErrorOccurred parseError: NSError) {
        print("error = \(parseError.localizedDescription)")
    }
    // beginParsing
    // This function begins the XML parsing process
    func beginParsing() {
        posts = []
        parser = NSXMLParser(contentsOfURL: NSURL(string: feed)!)!
        parser.delegate = self
        if (!parser.parse()) {
            print("parser failed")
            print("error is \(parser.parserError)")
        }
    }
    // didStartElement
    func parser(parser: NSXMLParser, didStartElement elementName: String, namespaceURI: String?, qualifiedName qName: String?, attributes attributeDict: [String : String]) {
        element = elementName
        if (element == "item") {
            articleTitle = ""
            articleLink = ""
        }
    }
    // foundCharacters
    func parser(parser: NSXMLParser, foundCharacters string: String) {
        if (element == "title") {
            articleTitle.appendString(string)
        } else if (element == "link") {
            articleLink.appendString(string)
        }
    }
    // didEndElement
    func parser(parser: NSXMLParser, didEndElement elementName: String, namespaceURI: String?, qualifiedName qName: String?) {
        if (elementName == "item") {
            if (!(articleTitle.isEqual(nil)) && !articleLink.isEqual(nil)) {
                let dictionaryEntry: Dictionary = ["title": (articleTitle as String), "link": (articleLink as String)]
                posts.append(dictionaryEntry)
            }
        }
    }
}